Gradle 8 compiled plugins are not usable with Gradle <7

### Expected Behavior
When I compile a plugin with the latest Gradle version, I expect it to work with older versions, otherwise it's not possible to support a range of Gradle versions. See e.g. Kotlin or Gradle Enterprise plugins which do this extensively.

### Current Behavior
```kotlin
abstract class SomeTask: DefaultTask() // not necessary to be defined here, just to show the inheritance tree

fun something(some: SomeTask) {
   some.inputs.files(...)
}
```
compiling the above against `implementation(gradleApi())` in Gradle 8 RC1, it binds (at compile time) to

https://github.com/gradle/gradle/blob/287aac28bc8f1de1b1b74852da8b254cc376bbfb/subprojects/core/src/main/java/org/gradle/api/internal/TaskInputsInternal.java#L31-L32

### Steps to Reproduce 
1. Compile a plugin with Gradle 8 which calls `task.inputs.files(...)`
2. Use this plugin in Gradle 7.x
3. Stand in awe of this beauty:
```
Caused by: java.lang.NoSuchMethodError: 'org.gradle.api.internal.tasks.TaskInputFilePropertyBuilderInternal org.gradle.api.internal.TaskInputsInternal.files(java.lang.Object[])'
	at com.example.SomethingKt.something(something.kt:4)
```

### Context
@lptr added these methods recently: https://github.com/gradle/gradle/pull/22739
It looks like we've been here before as @big-guy explains with Gradle 3.2 breaking compatibility: https://github.com/gradle/gradle/issues/1030
https://github.com/gradle/gradle/issues/1004 is the issue that's being regressed I guess.
Related, or potentially "caused by": https://github.com/gradle/gradle/issues/23431

### Workaround
`(some as Task).inputs` will force the compile-time binding to resolve `Task.getInputs(): TaskInputs` which doesn't have the `files()` overload and therefore compatible with all Gradle versions.

"Gradle Plugin author DX" will be disrupted if this stays like this.

### Your Environment
<!--- Include as many relevant details about the environment you experienced the bug in -->
<!--- A build scan https://scans.gradle.com/get-started is ideal -->
Build scan URL: it's simple to see the problem statically, but if really necessary I can set up a running example.
For real examples: [see this build's test report artifacts](https://github.com/TWiStErRob/net.twisterrob.gradle/actions/runs/3866626055) (tests in this project runs many Gradle builds for compatibility testing).

--- a/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskNodeCodec.kt
+++ b/subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/TaskNodeCodec.kt
@@ -24,6 +24,7 @@ import org.gradle.api.internal.TaskInternal
 import org.gradle.api.internal.TaskOutputsInternal
 import org.gradle.api.internal.provider.Providers
 import org.gradle.api.internal.tasks.TaskDestroyablesInternal
+import org.gradle.api.internal.tasks.TaskInputFilePropertyBuilderInternal
 import org.gradle.api.internal.tasks.TaskLocalStateInternal
 import org.gradle.api.internal.tasks.properties.InputParameterUtils
 import org.gradle.api.specs.Spec
@@ -421,13 +422,13 @@ suspend fun ReadContext.readInputPropertiesOf(task: Task) =
                     val normalizer = readEnum<InputNormalizer>()
                     val directorySensitivity = readEnum<DirectorySensitivity>()
                     val lineEndingNormalization = readEnum<LineEndingSensitivity>()
-                    (task as TaskInternal).inputs.run {
+                    ((task as TaskInternal).inputs.run {
                         when (filePropertyType) {
                             InputFilePropertyType.FILE -> file(pack(propertyValue))
                             InputFilePropertyType.DIRECTORY -> dir(pack(propertyValue))
                             InputFilePropertyType.FILES -> files(pack(propertyValue))
                         }
-                    }.run {
+                    } as TaskInputFilePropertyBuilderInternal).run {
                         withPropertyName(propertyName)
                         optional(optional)
                         skipWhenEmpty(inputBehavior.shouldSkipWhenEmpty())

--- a/subprojects/core/src/main/java/org/gradle/api/internal/TaskInputsInternal.java
+++ b/subprojects/core/src/main/java/org/gradle/api/internal/TaskInputsInternal.java
@@ -17,7 +17,6 @@
 package org.gradle.api.internal;
 
 import org.gradle.api.internal.tasks.TaskDependencyContainer;
-import org.gradle.api.internal.tasks.TaskInputFilePropertyBuilderInternal;
 import org.gradle.api.tasks.TaskInputs;
 import org.gradle.internal.properties.PropertyVisitor;
 
@@ -27,13 +26,4 @@ public interface TaskInputsInternal extends TaskInputs, TaskDependencyContainer 
      * Calls the corresponding visitor methods for all inputs added via the runtime API.
      */
     void visitRegisteredProperties(PropertyVisitor visitor);
-
-    @Override
-    TaskInputFilePropertyBuilderInternal files(Object... paths);
-
-    @Override
-    TaskInputFilePropertyBuilderInternal file(Object path);
-
-    @Override
-    TaskInputFilePropertyBuilderInternal dir(Object dirPath);
 }

--- a/subprojects/integ-test/src/crossVersionTest/groovy/org/gradle/integtests/AbstractTaskSubclassingBinaryCompatibilityCrossVersionSpec.groovy
+++ b/subprojects/integ-test/src/crossVersionTest/groovy/org/gradle/integtests/AbstractTaskSubclassingBinaryCompatibilityCrossVersionSpec.groovy
@@ -31,7 +31,8 @@ import org.gradle.api.tasks.compile.JavaCompile
 import org.gradle.api.tasks.scala.ScalaCompile
 import org.gradle.api.tasks.testing.Test
 import org.gradle.integtests.fixtures.CrossVersionIntegrationSpec
-import org.gradle.integtests.fixtures.TargetVersions
+import org.gradle.jvm.application.tasks.CreateStartScripts
+import org.gradle.jvm.tasks.Jar
 import org.gradle.plugins.ear.Ear
 import org.gradle.plugins.ide.eclipse.GenerateEclipseClasspath
 import org.gradle.plugins.ide.eclipse.GenerateEclipseJdt
@@ -45,13 +46,10 @@ import org.gradle.plugins.signing.Sign
 import org.gradle.util.GradleVersion
 
 /**
- * Tests that task classes compiled against earlier versions of Gradle are still compatible.
+ * Tests that task classes compiled against one version of Gradle are compatible with another version.
  */
-@TargetVersions("3.0+")
-class TaskSubclassingBinaryCompatibilityCrossVersionSpec extends CrossVersionIntegrationSpec {
-    @SuppressWarnings("UnnecessaryQualifiedReference")
-    def "can use task subclass compiled using previous Gradle version"() {
-        given:
+abstract class AbstractTaskSubclassingBinaryCompatibilityCrossVersionSpec extends CrossVersionIntegrationSpec {
+    protected void prepareSubclassingTest(GradleVersion targetVersion) {
         def taskClasses = [
             DefaultTask,
             SourceTask,
@@ -84,31 +82,31 @@ class TaskSubclassingBinaryCompatibilityCrossVersionSpec extends CrossVersionInt
 
         // Task types added after 1.0
 
-        if (previous.version >= GradleVersion.version("2.4")) {
-            taskClasses += org.gradle.jvm.application.tasks.CreateStartScripts
+        if (targetVersion >= GradleVersion.version("2.4")) {
+            taskClasses += CreateStartScripts
         }
-        if (previous.version >= GradleVersion.version("2.3")) {
-            taskClasses += org.gradle.jvm.tasks.Jar
+        if (targetVersion >= GradleVersion.version("2.3")) {
+            taskClasses += Jar
         }
 
         // Some breakages that were not detected prior to release. Please do not add any more exceptions
 
-        if (previous.version < GradleVersion.version("1.1")) {
+        if (targetVersion < GradleVersion.version("1.1")) {
             // Breaking changes were made to Test between 1.0 and 1.1
             taskClasses.remove(Test)
         }
-        if (previous.version < GradleVersion.version("2.0")) {
+        if (targetVersion < GradleVersion.version("2.0")) {
             // Breaking changes were made to JavaCompile prior to 2.0
             taskClasses.remove(JavaCompile)
         }
 
         Map<String, String> subclasses = taskClasses.collectEntries { ["custom" + it.name.replace(".", "_"), it.name] }
         def apiDepConf = "implementation"
-        if (previous.version < GradleVersion.version("7.0-rc-1")) {
+        if (targetVersion < GradleVersion.version("7.0-rc-1")) {
             apiDepConf = "compile"
         }
         def groovyDepConf
-        if (previous.version < GradleVersion.version("1.4-rc-1")) {
+        if (targetVersion < GradleVersion.version("1.4-rc-1")) {
             groovyDepConf = "groovy"
         } else {
             groovyDepConf = apiDepConf
@@ -126,14 +124,14 @@ class TaskSubclassingBinaryCompatibilityCrossVersionSpec extends CrossVersionInt
             import org.gradle.api.Project
 
             class SomePlugin implements Plugin<Project> {
-                void apply(Project p) { """ << \
-            subclasses.collect { "p.tasks.create('${it.key}', ${it.key})" }.join("\n") << """
+                void apply(Project p) { """ <<      \
+                 subclasses.collect { "p.tasks.create('${it.key}', ${it.key})" }.join("\n") << """
                 }
             }
-            """ << \
-            subclasses.collect {
-                def className = it.key
-                """class ${className} extends ${it.value} {
+            """ <<      \
+                 subclasses.collect {
+            def className = it.key
+            """${declareTaskClass(targetVersion)} class ${className} extends ${it.value} {
     ${className}() {
         // GRADLE-3185
         project.logger.lifecycle('task created')
@@ -141,7 +139,7 @@ class TaskSubclassingBinaryCompatibilityCrossVersionSpec extends CrossVersionInt
         super.getServices()
     }
 }"""
-            }.join("\n")
+        }.join("\n")
 
         buildFile << """
 buildscript {
@@ -150,25 +148,20 @@ buildscript {
 
 apply plugin: SomePlugin
 """
-
-        expect:
-        version previous withTasks 'assemble' inDirectory(file("producer")) run()
-        version current withTasks 'tasks' requireDaemon() requireIsolatedDaemons() run()
     }
 
-    def "task can use all methods declared by Task interface that AbstractTask specialises"() {
+    protected void prepareMethodUseTest(GradleVersion targetVersion) {
         file("someFile").touch()
         file("anotherFile").touch()
         file("yetAnotherFile").touch()
         file("someDir").createDir()
 
-        when:
         def apiDepConf = "implementation"
-        if (previous.version < GradleVersion.version("7.0-rc-1")) {
+        if (targetVersion < GradleVersion.version("7.0-rc-1")) {
             apiDepConf = "compile"
         }
         def groovyDepConf
-        if (previous.version < GradleVersion.version("1.4-rc-1")) {
+        if (targetVersion < GradleVersion.version("1.4-rc-1")) {
             groovyDepConf = "groovy"
         } else {
             groovyDepConf = apiDepConf
@@ -181,8 +174,8 @@ apply plugin: SomePlugin
             }
         """
 
-        boolean previousVersionLeaksInternal = (previous.version == GradleVersion.version("3.2") ||
-            previous.version == GradleVersion.version("3.2.1"))
+        boolean previousVersionLeaksInternal = (targetVersion == GradleVersion.version("3.2") ||
+            targetVersion == GradleVersion.version("3.2.1"))
 
         file("producer/src/main/java/SubclassTask.java") << """
             import org.gradle.api.DefaultTask;
@@ -190,14 +183,14 @@ apply plugin: SomePlugin
             import org.gradle.api.logging.LogLevel;
             import java.util.*;
 
-            public class SubclassTask extends DefaultTask {
+            public ${declareTaskClass(targetVersion)} class SubclassTask extends DefaultTask {
                 public SubclassTask() {
                     ${previousVersionLeaksInternal ? "((TaskInputs)getInputs())" : "getInputs()"}.file("someFile");
                     ${previousVersionLeaksInternal ? "((TaskInputs)getInputs())" : "getInputs()"}.files("anotherFile", "yetAnotherFile");
                     ${previousVersionLeaksInternal ? "((TaskInputs)getInputs())" : "getInputs()"}.dir("someDir");
-                    ${previous.version >= GradleVersion.version("4.3")
-                        ? 'getInputs().property("input", "value");'
-                        : ""}
+                    ${targetVersion >= GradleVersion.version("4.3")
+            ? 'getInputs().property("input", "value");'
+            : ""}
                     Map<String, Object> mapValues = new HashMap<String, Object>();
                     mapValues.put("mapInput", "mapValue");
                     ${previousVersionLeaksInternal ? "((TaskInputs)getInputs())" : "getInputs()"}.properties(mapValues);
@@ -224,9 +217,13 @@ apply plugin: SomePlugin
 
             task t(type: SubclassTask)
         """
+    }
 
-        then:
-        version previous withTasks 'assemble' inDirectory(file("producer")) run()
-        version current requireDaemon() requireIsolatedDaemons() withTasks 't' run()
+    protected static String declareTaskClass(GradleVersion targetVersion) {
+        if (targetVersion >= GradleVersion.version("7.0")) {
+            return "abstract"
+        } else {
+            return ""
+        }
     }
 }

--- a/subprojects/integ-test/src/crossVersionTest/groovy/org/gradle/integtests/TaskSubclassingBinaryBackwardsCompatibilityCrossVersionSpec.groovy
+++ b/subprojects/integ-test/src/crossVersionTest/groovy/org/gradle/integtests/TaskSubclassingBinaryBackwardsCompatibilityCrossVersionSpec.groovy
@@ -1,0 +1,42 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.integtests
+
+import org.gradle.integtests.fixtures.TargetVersions
+
+/**
+ * Tests that task classes compiled with the current version of Gradle are compatible with previous versions.
+ */
+@TargetVersions("7.0+")
+class TaskSubclassingBinaryBackwardsCompatibilityCrossVersionSpec extends AbstractTaskSubclassingBinaryCompatibilityCrossVersionSpec {
+    def "can use task subclass using previous Gradle version"() {
+        given:
+        prepareSubclassingTest(current.version)
+
+        expect:
+        version current withTasks 'assemble' inDirectory(file("producer")) run()
+        version previous withTasks 'tasks' requireDaemon() requireIsolatedDaemons() run()
+    }
+
+    def "task can use all methods declared by Task interface that AbstractTask specialises"() {
+        given:
+        prepareMethodUseTest(current.version)
+
+        expect:
+        version current withTasks 'assemble' inDirectory(file("producer")) run()
+        version previous requireDaemon() requireIsolatedDaemons() withTasks 't' run()
+    }
+}

--- a/subprojects/integ-test/src/crossVersionTest/groovy/org/gradle/integtests/TaskSubclassingBinaryForwardCompatibilityCrossVersionSpec.groovy
+++ b/subprojects/integ-test/src/crossVersionTest/groovy/org/gradle/integtests/TaskSubclassingBinaryForwardCompatibilityCrossVersionSpec.groovy
@@ -1,0 +1,42 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.gradle.integtests
+
+import org.gradle.integtests.fixtures.TargetVersions
+
+/**
+ * Tests that task classes compiled against earlier versions of Gradle are still compatible.
+ */
+@TargetVersions("3.0+")
+class TaskSubclassingBinaryForwardCompatibilityCrossVersionSpec extends AbstractTaskSubclassingBinaryCompatibilityCrossVersionSpec {
+    def "can use task subclass compiled using previous Gradle version"() {
+        given:
+        prepareSubclassingTest(previous.version)
+
+        expect:
+        version previous withTasks 'assemble' inDirectory(file("producer")) run()
+        version current withTasks 'tasks' requireDaemon() requireIsolatedDaemons() run()
+    }
+
+    def "task can use all methods declared by Task interface that AbstractTask specialises"() {
+        given:
+        prepareMethodUseTest(previous.version)
+
+        expect:
+        version previous withTasks 'assemble' inDirectory(file("producer")) run()
+        version current requireDaemon() requireIsolatedDaemons() withTasks 't' run()
+    }
+}

